공공 데이터 API에서 데이터를 조회한다.

this-is-spear opened the pull request:
## 작업에 대한 간단한 요약

공공 데이터에서 필요한 데이터를 추출한다.

### 변경내역

- 공공 데이터에서 데이터를 조회한다.
- 공공 데이터에서 제공하는 데이터 10개를 조회하는 테스트를 구성한다.

### 어떤부분에 집중해서 리뷰하면좋을지

#### 서비스 키 관리 이슈

서비스 키를 외부에서 관리해야 하지만 필드에 들어간 문제가 있다.

#### 공공 데이터 API 필드 관리

유사한 이름의 필드 값을 가져올 때 코드 실수가 최대한 없앨 수 있도록 ENUM 을 활용할 필요가 있었다.

```java
public enum FoodNtrIrdntInfoField {  
  NAME("DESC_KOR"),  
  SERVING_SIZE_GRAM("SERVING_WT"),  
  CALORIES_KCAL("NUTR_CONT1"),  
  CARBOHYDRATES_GRAM("NUTR_CONT2"),

  ...

}
```

**AS-IS**

```java
return new FoodNtrIrdntInfoResponseItem(  
	node.get("DESC_KOR").asText(),  
	parseDouble(node.get("SERVING_WT")),  
	parseDouble(node.get("NUTR_CONT1")),  
	parseDouble(node.get("NUTR_CONT2")),  
	parseDouble(node.get("NUTR_CONT3")),  
	parseDouble(node.get("NUTR_CONT4")),  
	parseDouble(node.get("NUTR_CONT5")),  
	parseDouble(node.get("NUTR_CONT6")),  
	parseDouble(node.get("NUTR_CONT7")),  
	parseDouble(node.get("NUTR_CONT8")),  
	parseDouble(node.get("NUTR_CONT9")),  
	parseYear(node.get("BGN_YEAR")),  
	node.get("ANIMAL_PLANT").asText()
);
```

**TO-BE**

어떤 필드에 어떤 값이 들어가는지를 쉽게 확인해 휴먼 에러를 최소화했다.

```java
return new FoodNtrIrdntInfoResponseItem(  
	getField(node, NAME).asText(),  
	parseDouble(getField(node, SERVING_SIZE_GRAM)),  
	parseDouble(getField(node, CALORIES_KCAL)),  
	parseDouble(getField(node, CARBOHYDRATES_GRAM)),  
	parseDouble(getField(node, PROTEIN_GRAM)),  
	parseDouble(getField(node, FAT_GRAM)),  
	parseDouble(getField(node, SUGARS_GRAM)),  
	parseDouble(getField(node, SODIUM_MILLI_GRAM)),  
	parseDouble(getField(node, CHOLESTRAOL_MILLI_GRAM)),  
	parseDouble(getField(node, SATURATED_FATTY_ACIDS_GRAM)),  
	parseDouble(getField(node, TRANS_FATTY_ACIDS_GRAM)),  
	parseYear(getField(node, BEGIN_YEAR)),  
	getField(node, MANUFACTURER).asText()  
);
```

#### `JsonDeserializer`를 구현한 직접 파싱

응답 값 데이터 무결성이 지켜지지 않아 Double 자료구조 역직렬화가 진행되지 않았고, `N/A` 필드를 null 로 변경할 수 있는 수단이 필요했다.

**CASE 1**

```json
"NUTR_CONT7": "N/A"
```

**CASE 2**

```json
"NUTR_CONT7": "0.00"
```

내린 결론은 직접 파싱하는 일이었다.

```java
private Double parseDouble(JsonNode node) {  
	if ("N/A".equals(node.asText())) {  
		return null;  
	}  
	return node.asDouble();  
}
```

다음 코드에서 각 필드마다 검증하고 있다.

```java
public class CustomItemDeserializer extends JsonDeserializer<FoodNtrIrdntInfoResponseItem> {  
	@Override  
	public FoodNtrIrdntInfoResponseItem deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {  
		var node = (JsonNode) p.getCodec().readTree(p);  
		return new FoodNtrIrdntInfoResponseItem(  
			...
			parseDouble(getField(node, SERVING_SIZE_GRAM)),  
			...
		);  
	}
...
}
```

### 이슈번호 첨부,  기타내용

- [Projects - 공공데이터에서 데이터를 추출한다.](https://github.com/orgs/towards-sweetly/projects/1?pane=issue&itemId=33293668)

this-is-spear committed:
feature : 데이터 조회 기능 추가 및 테스트

youngreal commented on src/test/java/towardssweetly/datascheduler/infra/FoodNtrIrdntInfoServiceTest.java:
```diff
@@ -0,0 +1,21 @@
+        final String json = "json";
+        final int pageNo = 1;
+        final int numOfRows = 10;
+        var response = foodNtrIrdntInfoService.getData(pageNo, numOfRows, json, SERVICE_KEY);
```
별건아니지만 이부분은 반환타입이 명확하지않아서 저같은경우 getData메서드를 한번더 들어가서 확인했는데
저는 반환타입을 다른위치에서 호출하는 경우에는 var보단 그냥 명시적타입을 그냥 사용하는게 나을수도 있다는 생각이 들긴합니다ㅠ (final이 빠진것같네요??)

this-is-spear replied:
IDE에서 적절하게 힌트를 제공해줘서 필요하지 않겠다는 생각을 했습니다! 

<img width="989" alt="스크린샷 2023-08-02 오전 9 58 41" src="https://github.com/towards-sweetly/data-scheduler/assets/92219795/c871309e-ace0-4aed-9a21-46ecd061802d">

반환값을 유추할 수 없었던 이유는 `getData()`라는 모호한 메서드명 때문이라 생각듭니다. `getData()`에서 `getFoodNtrIrdntInfoResponse(...)`로 메서드 네이밍 변경하겠습니다.

youngreal commented on src/main/java/towardssweetly/datascheduler/infra/dto/FoodNtrIrdntInfoResponse.java:
```diff
@@ -0,0 +1,14 @@
+@Getter
+@ToString
+public class FoodNtrIrdntInfoResponse {
+    @JsonProperty("header")
+    private FoodNtrIrdntInfoResponseHeader header;
+    @JsonProperty("body")
+    private FoodNtrIrdntInfoResponseBody body;
```
저는 자바 코드와 json으로 내려가는 키이름이 다른경우에만 해당 어노테이션으로 맞춰줬었는데 그냥 모든필드에 작성하기로 하신건가요? 아니면 다른이점이 있는건가요?

this-is-spear replied:
필드 명 변경 가능성을 염두했는데, 오히려 관리할 부분만 늘어나게 되는군요. 이 부분은 필요한 필드만 `@JsonProperty` 애너테이션을 추가하겠습니다.

youngreal reviewed (approved):
고생하셨습니다~

this-is-spear committed:
refact : 메서드 네이밍 변경

###Description###

- getData -> getFoodNtrIrdntInfoResponse로 메서드 네이밍 변경
- getDataTest -> getFoodNtrIrdntInfoResponseTest 메서드 네이밍 변경
- getFoodNtrIrdntInfoResponseTest 블록 변수 타입 var로 변경
- getFoodNtrIrdntInfoResponseTest 블록 변수 불변 설정

this-is-spear committed:
refact : 동일한 이름의 json property 제거
